test/api_server_unittest.cc

```diff
@@ -0,0 +1,89 @@
+#include "../src/api_server.h"
+#include "../src/configuration.h"
+#include "../src/store.h"
+#include "gtest/gtest.h"
+
+namespace google {
+
+class ApiServerTest : public ::testing::Test {
+ protected:
+  static void BasicDispatcher() {
+    bool handler_called;
+    MetadataApiServer::Handler handler = [&handler_called](
+        const MetadataApiServer::HttpServer::request& request,
+        std::shared_ptr<MetadataApiServer::HttpServer::connection> conn) {
+      handler_called = true;
+    };
+    MetadataApiServer::Handler badHandler = [&handler_called](
+        const MetadataApiServer::HttpServer::request& request,
+        std::shared_ptr<MetadataApiServer::HttpServer::connection> conn) {
+      handler_called = false;
+    };
+    MetadataApiServer::Dispatcher::HandlerMap handlers({
+      {{"GET", "/testPath/"}, handler},
+      {{"GET", "/badPath/"}, badHandler},
+    });
+    MetadataApiServer::Dispatcher dispatcher(handlers, false);
+    MetadataApiServer::HttpServer::request request;
+    request.method = "GET";
+    request.destination = "/testPath/";
+    dispatcher(request, nullptr);
+    EXPECT_TRUE(handler_called);
+  }
+
+  static void DispatcherMethodCheck() {
+    bool handler_called;
+    MetadataApiServer::Handler handler = [&handler_called](
+        const MetadataApiServer::HttpServer::request& request,
+        std::shared_ptr<MetadataApiServer::HttpServer::connection> conn) {
+      handler_called = true;
+    };
+    MetadataApiServer::Dispatcher::HandlerMap handlers({
+      {{"GET", "/testPath/"}, handler},
+    });
+    MetadataApiServer::Dispatcher dispatcher(handlers, false);
+    MetadataApiServer::HttpServer::request request;
+    request.method = "POST";
+    request.destination = "/testPath/";
+    dispatcher(request, nullptr);
+    EXPECT_FALSE(handler_called);
+  }
+
+  static void DispatcherSubstringCheck() {
+    bool handler_called;
+    MetadataApiServer::Handler handler = [&handler_called](
+        const MetadataApiServer::HttpServer::request& request,
+        std::shared_ptr<MetadataApiServer::HttpServer::connection> conn) {
+      handler_called = true;
+    };
+    MetadataApiServer::Dispatcher::HandlerMap handlers({
+      {{"GET", "/testPath/"}, handler},
+    });
+    MetadataApiServer::Dispatcher dispatcher(handlers, false);
+    MetadataApiServer::HttpServer::request request;
+    request.method = "GET";
+    request.destination = "/testPathFoo/";
+    dispatcher(request, nullptr);
+    EXPECT_FALSE(handler_called);
+    request.destination = "/test/";
+    dispatcher(request, nullptr);
+    EXPECT_FALSE(handler_called);
+    request.destination = "/testFooPath/";
+    dispatcher(request, nullptr);
+    EXPECT_FALSE(handler_called);
+  }
+};
+
+TEST_F(ApiServerTest, BasicDispacher) {
+  BasicDispatcher();
```

Comment:
Hmm... Do you have to put the whole test into a static function in the fixture? Would it make more sense to add functions to create and invoke the dispatcher instead? Also below.

Reply:
The problem is that HttpServer, Handler, HandlerMap, and even Dispatcher, are all declared private.  We could attempt to pull out small portions of the test, but that seems confusing and difficult without any gain.

Reply:
Very cool, changed, much better.